infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/receive/StreamsSnapshotWriter.java

```diff
@@ -0,0 +1,273 @@
+     * @param currentSeqNum
+     * @param dstUUID
+     */
+    void processOpaqueEntry(List<SMREntry> smrEntries, Long currentSeqNum, UUID dstUUID) {
+        CorfuStoreMetadata.Timestamp timestamp = persistedWriterMetadata.getTimestamp();
+        long persistConfigID = persistedWriterMetadata.query(timestamp, PersistedWriterMetadata.PersistedWriterMetadataType.SiteConfigID);
+        long persistSnapStart = persistedWriterMetadata.query(timestamp, PersistedWriterMetadata.PersistedWriterMetadataType.LastSnapStart);
+        long persitSeqNum = persistedWriterMetadata.query(timestamp, PersistedWriterMetadata.PersistedWriterMetadataType.LastSnapSeqNum);
+
+        if (siteConfigID != persistConfigID || srcGlobalSnapshot != persistSnapStart || currentSeqNum != (persitSeqNum + 1)) {
+            log.warn("Skip current siteConfigID " + siteConfigID + " srcGlobalSnapshot " + srcGlobalSnapshot + " currentSeqNum " + currentSeqNum +
+                    " persistedMetadata " + persistedWriterMetadata.getSiteConfigID() + " startSnapshot " + persistedWriterMetadata.getLastSnapStartTimestamp() +
+                    " lastSnapSeqNum " + persistedWriterMetadata.getLastSnapSeqNum());
+            return;
+        }
+
+        TxBuilder txBuilder = persistedWriterMetadata.getTxBuilder();
+        persistedWriterMetadata.appendUpdate(txBuilder, PersistedWriterMetadata.PersistedWriterMetadataType.SiteConfigID, siteConfigID);
+        persistedWriterMetadata.appendUpdate(txBuilder, PersistedWriterMetadata.PersistedWriterMetadataType.LastSnapStart, srcGlobalSnapshot);
+        persistedWriterMetadata.appendUpdate(txBuilder, PersistedWriterMetadata.PersistedWriterMetadataType.LastSnapSeqNum, currentSeqNum);
+        for (SMREntry smrEntry : smrEntries) {
+            txBuilder.logUpdate(dstUUID, smrEntry);
+        }
+
+        try {
+            txBuilder.commit(timestamp);
+        } catch (Exception e) {
+            log.warn("Caught an exception ", e);
+            throw e;
+        }
+        log.debug("Process the entries {}  and set sequence number {} ", smrEntries, currentSeqNum);
+    }
+
+    @Override
+    public void apply(LogReplicationEntry message) {
+        verifyMetadata(message.getMetadata());
+
+        if (message.getMetadata().getSnapshotSyncSeqNum() != recvSeq ||
+                message.getMetadata().getMessageMetadataType() != MessageType.SNAPSHOT_MESSAGE) {
+            log.error("Expecting sequencer {} != recvSeq {} or wrong message type {} expecting {}",
+                    message.getMetadata().getSnapshotSyncSeqNum(), recvSeq,
+                    message.getMetadata().getMessageMetadataType(), MessageType.SNAPSHOT_MESSAGE);
+            throw new ReplicationWriterException("Message is out of order or wrong type");
+        }
+
+        byte[] payload = message.getPayload();
+        OpaqueEntry opaqueEntry = OpaqueEntry.deserialize(Unpooled.wrappedBuffer(payload));
+
+        if (opaqueEntry.getEntries().keySet().size() != 1) {
```

Comment:
what does this mean. we will return silently if entries is not exactly 1. can this not lead to bugs ?